sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,539 @@
+# Dynamic linking of device code
+
+This document describes purpose and design of dynamic linking of device code
+feature.
+
+## Background
+Sometimes users want to link device code dynamically at run time. One possible
+use case for such linkage - providing device functions via shared libraries.
+The example below shows how device function `LibDeviceFunc` can be dynamically
+linked to a SYCL app:
+```
+// app.cpp
+SYCL_EXTERNAL int LibDeviceFunc(int i);
+class KernelName;
+/* ... */
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for<KernelName>(/* ... */ [=](sycl::item i) {
+  out[i] = LibDeviceFunc(i);
+}); /* ... */
+std::cout << out[i] << “ “;
+
+// lib.cpp
+int SYCL_EXTERNAL LibDeviceFunc(int i) {
+  return i * 2;
+}
+
+// Commands
+clang++ -fsycl lib.cpp -shared -o helpers.so
+clang++ -fsycl app.cpp -lhelpers -o a.out
+./a.out
+Output: 0 2 4 6…
+
+```
+It is possible to manually create `sycl::program` in both app and shared
+library, then use `link` SYCL API to get a single program and launch kernels
+using it. But it is not user-friendly and it is very different from regular
+C/C++ workflow.
+
+Another possible scenario - use functions defined in a pre-compiled device image
+provided by the user. Example:
+```
+// a.cpp
+SYCL_EXTERNAL void LibDeviceFunc();
+...
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([]() { LibDeviceFunc(); });
+});
+
+// b.cpp
+/*no SYCL_EXTERNAL*/ void LibDeviceFunc() { ... }
+```
+We have a `SYCL_EXTERNAL` function `foo` called from a kernel, but the
+application defined only host version of this function. Then user adds device
+image with definition of `foo` to the fat object via special compiler option
+(like `-fsycl-add-targets`).
+
+The main purpose of this feature is to provide a mechanism which allows to
+link device code dynamically at runtime, such as in the scenarios above.
+
+## Requirements:
+User's device code that consists of some device API (`SYCL_EXTERNAL` functions),
+is compiled into some form and it is not linked statically with device code of
+application. It can be a shared library that contains some device code or a
+separate device image supplied with properties attached. This code is linked
+dynamically at run time with device code of a user's application in order to
+resolve dependencies.
+For this combination the following statements must be true:
+
+The presented dynamic device code linkage mechanism must:
+
+- Allow to represent the actual dynamically linked code as a device binary image
+  which can be:
+  - Embedded into a host shared object by standard SYCL compiler driver
+    invocation
+  - Embedded into a host binary or shared object using manual invocations of
+    SYCL tools such as `clang-offload-wrapper` and linker
+  - Loaded into memory via special API
```

Comment:
> Loaded into memory via special API

Into host memory from a file? Or something else? What is the API (SYCL, PI, anything else) envisioned?

Reply:
This is idea for a new API proposed by @kbobrovs in previous comments, see https://github.com/intel/llvm/pull/3210#discussion_r613537978 .
I guess, the device image will be loaded into host memory from a file and this will be some SYCL API.

Reply:
I see now. There is not enough info in what you have added. Also @romanovvlad is suggesting that standard dlopen could be made to work for that. If you want to leave this TODO then please describe the requirement more clearly.

Reply:
I don't have strong objections regarding such API since it can be more convenient than creation of (empty?) shared host object and loading via dlopen. I think we need to hear opinion regarding this suggestion from @kbobrovs since he originally proposed new API.